File: jogoCobrinha.py
#!/usr/bin/python3

import os
import sys
import pygame
import random
from fcntl import ioctl

# Mapeamento dos dígitos para os valores correspondentes no display de 7 segmentos
SEGMENT_MAP = {
    0: 0b11000000,
    1: 0b11111001,
    2: 0b10100100,
    3: 0b10110000,
    4: 0b10011001,
    5: 0b10010010,
    6: 0b10000010,
    7: 0b11111000,
    8: 0b10000000,
    9: 0b10010000
}

def encode_number(num: int, fd):
    high, low = 0, 0

    for i in range(8):
        digit = num % 10  # Obtém o último dígito
        num //= 10  # Remove o último dígito
        
        segment_value = SEGMENT_MAP[digit]  # Obtém o valor para 7 segmentos

        if i == 0:
        	low |= (segment_value << 0)
        elif i < 4:
        	low |= (segment_value << (i * 8))
        else:
        	high |= (segment_value << ((i - 4) * 8)) 
        	
        ioctl(fd, WR_L_DISPLAY)
        retval = os.write(fd, low.to_bytes(4, 'little'))
        ioctl(fd, WR_R_DISPLAY)
        retval = os.write(fd, high.to_bytes(4, 'little'))
        
# ioctl commands defined at the pci driver
RD_SWITCHES   = 24929
RD_PBUTTONS   = 24930
WR_L_DISPLAY  = 24931
WR_R_DISPLAY  = 24932
WR_RED_LEDS   = 24933
WR_GREEN_LEDS = 24934

# Inicialização do pygame
pygame.init()

# Configurações da tela
WIDTH, HEIGHT = 600, 400
BLOCK_SIZE = 20
screen = pygame.display.set_mode((WIDTH, HEIGHT))
pygame.display.set_caption('Jogo da Cobrinha')

# Definição de cores
WHITE = (255, 255, 255)
BLACK = (0, 0, 0)
RED   = (255, 0, 0)
GREEN = (0, 255, 0)

clock = pygame.time.Clock()

# Fonte para mensagens
font_style = pygame.font.SysFont(None, 35)

def message(msg, color, pos):
    mesg = font_style.render(msg, True, color)
    screen.blit(mesg, pos)

def draw_snake(snake_body):
    for block in snake_body:
        pygame.draw.rect(screen, GREEN, [block[0], block[1], BLOCK_SIZE, BLOCK_SIZE])

def new_food(snake_body):
    while True:
        x = random.randrange(0, WIDTH, BLOCK_SIZE)
        y = random.randrange(0, HEIGHT, BLOCK_SIZE)
        if [x, y] not in snake_body:
            return [x, y]

def gameLoop(fd):
    game_over = False
    game_close = False
    x = WIDTH // 2
    y = HEIGHT // 2
    x_change = 0
    y_change = 0
    snake_body = []
    snake_length = 1
    food = new_food(snake_body)
    encode_number(0, fd)

    while not game_over:
        while game_close:
            screen.fill(BLACK)
            message("Você perdeu! Pressione C para jogar ou Q para sair", RED, [0, HEIGHT / 2])
            pygame.display.update()
            ioctl(fd, RD_SWITCHES)
            switches = os.read(fd, 1)
            switches = int.from_bytes(switches, 'little')
            
            #if event.type == pygame.KEYDOWN:
            # sair
            print(switches)
            if switches & 0x1:
                game_over = True
                game_close = False
            if switches & 0x2:
                gameLoop(fd)

            for event in pygame.event.get():
                if event.type == pygame.QUIT:
                    game_over = True
                    game_close = False

        # Leitura dos botões
        ioctl(fd, RD_PBUTTONS)
        buttons = os.read(fd, 1)
        buttons = int.from_bytes(buttons, 'little')
		
        buttons = (~buttons)
        print(buttons)
        if ( buttons) & 0x2:  # Botão para esquerda
            x_change = -BLOCK_SIZE
            y_change = 0
        elif ( buttons) & 0x01:  # Botão para direita
            x_change = BLOCK_SIZE
            y_change = 0
        elif ( buttons) & 0x08:  # Botão para cima
            y_change = -BLOCK_SIZE
            x_change = 0
        elif ( buttons) & 0x04:  # Botão para baixo
            y_change = BLOCK_SIZE
            x_change = 0

        if x >= WIDTH or x < 0 or y >= HEIGHT or y < 0:
            game_close = True
        x += x_change
        y += y_change
        screen.fill(BLACK)
        pygame.draw.rect(screen, RED, [food[0], food[1], BLOCK_SIZE, BLOCK_SIZE])

        snake_head = [x, y]
        snake_body.append(snake_head)
        if len(snake_body) > snake_length:
            del snake_body[0]

        for block in snake_body[:-1]:
            if block == snake_head:
                game_close = True

        draw_snake(snake_body)
        pygame.display.update()

        if x == food[0] and y == food[1]:
            food = new_food(snake_body)
            snake_length += 1
            # Testando a função
            encode_number(snake_length-1, fd)

        clock.tick(5)

    pygame.quit()
    sys.exit()

def main():
    if len(sys.argv) < 2:
        print("Error: expected more command line arguments")
        print("Syntax: %s </dev/device_file>" % sys.argv[0])
        exit(1)

    fd = os.open(sys.argv[1], os.O_RDWR)
    gameLoop(fd)
    os.close(fd)

if __name__ == '__main__':
    main()

File: jogoCobrinha.cpp
#include <iostream>
#include <vector>
#include <cstdlib>
#include <ctime>
#include <unistd.h>
#include <fcntl.h>
#include <sys/ioctl.h>
#include <SDL2/SDL.h>

// ioctl commands defined at the PCI driver
#define RD_SWITCHES   24929
#define RD_PBUTTONS   24930
#define WR_L_DISPLAY  24931
#define WR_R_DISPLAY  24932
#define WR_RED_LEDS   24933
#define WR_GREEN_LEDS 24934

const int WIDTH = 600;
const int HEIGHT = 400;
const int BLOCK_SIZE = 20;
const SDL_Color WHITE = {255, 255, 255, 255};
const SDL_Color BLACK = {0, 0, 0, 255};
const SDL_Color RED = {255, 0, 0, 255};
const SDL_Color GREEN = {0, 255, 0, 255};

void message(SDL_Renderer* renderer, TTF_Font* font, const char* msg, SDL_Color color, SDL_Point pos) {
    SDL_Surface* surface = TTF_RenderText_Solid(font, msg, color);
    SDL_Texture* texture = SDL_CreateTextureFromSurface(renderer, surface);
    SDL_Rect dstrect = {pos.x, pos.y, surface->w, surface->h};
    SDL_RenderCopy(renderer, texture, NULL, &dstrect);
    SDL_DestroyTexture(texture);
    SDL_FreeSurface(surface);
}

void draw_snake(SDL_Renderer* renderer, const std::vector<SDL_Point>& snake_body) {
    SDL_SetRenderDrawColor(renderer, GREEN.r, GREEN.g, GREEN.b, GREEN.a);
    for (const auto& block : snake_body) {
        SDL_Rect rect = {block.x, block.y, BLOCK_SIZE, BLOCK_SIZE};
        SDL_RenderFillRect(renderer, &rect);
    }
}

SDL_Point new_food(const std::vector<SDL_Point>& snake_body) {
    SDL_Point food;
    bool valid_position = false;
    while (!valid_position) {
        food.x = (rand() % (WIDTH / BLOCK_SIZE)) * BLOCK_SIZE;
        food.y = (rand() % (HEIGHT / BLOCK_SIZE)) * BLOCK_SIZE;
        valid_position = true;
        for (const auto& block : snake_body) {
            if (block.x == food.x && block.y == food.y) {
                valid_position = false;
                break;
            }
        }
    }
    return food;
}

void game_loop(int fd) {
    SDL_Window* window = SDL_CreateWindow("Snake Game", SDL_WINDOWPOS_CENTERED, SDL_WINDOWPOS_CENTERED, WIDTH, HEIGHT, SDL_WINDOW_SHOWN);
    SDL_Renderer* renderer = SDL_CreateRenderer(window, -1, SDL_RENDERER_ACCELERATED);
    TTF_Font* font = TTF_OpenFont("arial.ttf", 35);

    bool game_over = false;
    bool game_close = false;
    int x = WIDTH / 2;
    int y = HEIGHT / 2;
    int x_change = 0;
    int y_change = 0;
    std::vector<SDL_Point> snake_body;
    int snake_length = 1;
    SDL_Point food = new_food(snake_body);

    while (!game_over) {
        while (game_close) {
            SDL_SetRenderDrawColor(renderer, BLACK.r, BLACK.g, BLACK.b, BLACK.a);
            SDL_RenderClear(renderer);
            message(renderer, font, "You lost! Press C to play or Q to quit", RED, {0, HEIGHT / 2});
            SDL_RenderPresent(renderer);

            ioctl(fd, RD_SWITCHES);
            char switches_char;
            read(fd, &switches_char, 1);
            int switches = static_cast<int>(switches_char);
            
            if (switches & 0x1) {
                game_over = true;
                game_close = false;
            }
            if (switches & 0x2) {
                game_loop(fd);
            }

            SDL_Event event;
            while (SDL_PollEvent(&event)) {
                if (event.type == SDL_QUIT) {
                    game_over = true;
                    game_close = false;
                }
            }
        }

        ioctl(fd, RD_PBUTTONS);
        char buttons_char;
        read(fd, &buttons_char, 1);
        int buttons = static_cast<int>(buttons_char);
        buttons = ~buttons;

        if (buttons & 0x2) {
            x_change = -BLOCK_SIZE;
            y_change = 0;
        } else if (buttons & 0x1) {
            x_change = BLOCK_SIZE;
            y_change = 0;
        } else if (buttons & 0x8) {
            y_change = -BLOCK_SIZE;
            x_change = 0;
        } else if (buttons & 0x4) {
            y_change = BLOCK_SIZE;
            x_change = 0;
        }

        if (x >= WIDTH || x < 0 || y >= HEIGHT || y < 0) {
            game_close = true;
        }
        x += x_change;
        y += y_change;

        SDL_SetRenderDrawColor(renderer, BLACK.r, BLACK.g, BLACK.b, BLACK.a);
        SDL_RenderClear(renderer);
        SDL_SetRenderDrawColor(renderer, RED.r, RED.g, RED.b, RED.a);
        SDL_Rect food_rect = {food.x, food.y, BLOCK_SIZE, BLOCK_SIZE};
        SDL_RenderFillRect(renderer, &food_rect);

        SDL_Point snake_head = {x, y};
        snake_body.push_back(snake_head);
        if (snake_body.size() > snake_length) {
            snake_body.erase(snake_body.begin());
        }

        for (size_t i = 0; i < snake_body.size() - 1; ++i) {
            if (snake_body[i].x == snake_head.x && snake_body[i].y == snake_head.y) {
                game_close = true;
                break;
            }
        }

        draw_snake(renderer, snake_body);
        SDL_RenderPresent(renderer);

        if (x == food.x && y == food.y) {
            food = new_food(snake_body);
            ++snake_length;
        }

        SDL_Delay(200);
    }

    TTF_CloseFont(font);
    SDL_DestroyRenderer(renderer);
    SDL_DestroyWindow(window);
    SDL_Quit();
}

int main(int argc, char* argv[]) {
    if (argc < 2) {
        std::cerr << "Error: expected more command line arguments" << std::endl;
        std::cerr << "Syntax: " << argv[0] << " </dev/device_file>" << std::endl;
        return 1;
    }

    int fd = open(argv[1], O_RDWR);
    if (fd < 0) {
        std::cerr << "Error: could not open device file" << std::endl;
        return 1;
    }

    srand(time(NULL));
    if (SDL_Init(SDL_INIT_VIDEO) < 0) {
        std::cerr << "Error: could not initialize SDL" << std::endl;
        return 1;
    }
    if (TTF_Init() == -1) {
        std::cerr << "Error: could not initialize SDL_ttf" << std::endl;
        return 1;
    }

    game_loop(fd);

    close(fd);
    TTF_Quit();
    SDL_Quit();
    return 0;
}
